File: utilslib/src/main/java/com/ashlikun/utils/AppUtils.java
package com.ashlikun.utils;

import android.app.Application;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.support.v4.content.FileProvider;
import android.text.TextUtils;

import com.ashlikun.utils.main.ProcessUtils;
import com.ashlikun.utils.provider.BaseContentProvider;

import java.io.File;

/**
 * 作者　　: 李坤
 * 创建时间:2016/12/30　9:59
 * 邮箱　　：496546144@qq.com
 * <p>
 * 功能介绍：
 */
public class AppUtils {
    private static boolean isDebug;
    private static Application myApp;
    private static int versionCode = -10088;
    private static String versionName;
    private static String packageName;
    public static ApplicationListener activityLifecycleCallbacks = new ApplicationListener();


    /**
     * 作者　　: 李坤
     * 创建时间: 2017/8/7 10:29
     * 邮箱　　：496546144@qq.com
     * <p>
     * 方法功能：一定要在Application里面调用
     */

    public static void init(Application myApp) {
        AppUtils.myApp = myApp;
        myApp.registerActivityLifecycleCallbacks(activityLifecycleCallbacks);
        //设置内容提供者的名字
        BaseContentProvider.AUTHORITY = myApp.getApplicationInfo().packageName + "ContentProvider";
    }

    public static void setDebug(boolean isDebug) {
        AppUtils.isDebug = isDebug;
    }


    public static Application getApp() {
        return AppUtils.myApp;
    }

    public static boolean isDebug() {
        return AppUtils.isDebug;
    }

    /**
     * 前后台监听
     *
     * @param listener
     */
    public static void addAppRunChang(ApplicationListener.OnChangListener listener) {
        activityLifecycleCallbacks.addOnChangListener(listener);
    }

    /**
     * 前后台监听
     *
     * @param listener
     */
    public static void removeAppRunChang(ApplicationListener.OnChangListener listener) {
        activityLifecycleCallbacks.removeOnChangListener(listener);
    }

    /**
     * 是否是主进程
     * 如果getCurProcessName为空就默认认为不是主进程
     *
     * @return
     */
    public static boolean isMainProcess() {
        String name = ProcessUtils.getCurProcessName();
        if (TextUtils.isEmpty(name)) {
            return false;
        }
        return name.equals(myApp.getPackageName());
    }

    /**
     * 是否APP进程
     * com.ogow.activity
     * com.ogow.activity:webview
     *
     * @return 获取失败也是APP进程
     */
    public static boolean isAppProcess() {
        String name = ProcessUtils.getCurProcessName();
        if (TextUtils.isEmpty(name)) {
            return true;
        }
        return name.contains(myApp.getPackageName());
    }

    /**
     * 兼容高版本的fileprovider
     *
     * @param file
     * @return
     */
    public static Uri getUri(File file) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            String authority = getApp().getPackageName() + ".Provider";
            return FileProvider.getUriForFile(getApp(), authority, file);
        } else {
            return Uri.fromFile(file);
        }
    }


    /**
     * 获取版本名称
     *
     * @return
     */
    public static String getVersionName() {
        if (TextUtils.isEmpty(versionName)) {
            PackageManager pm = AppUtils.getApp().getPackageManager();
            try {
                PackageInfo packageInfo = pm.getPackageInfo(AppUtils.getApp().getPackageName(), 0);
                versionName = packageInfo.versionName;
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
        if (versionName == null) {
            versionName = "";
        }
        return versionName;
    }

    /**
     * 获取版本code
     *
     * @return
     */
    public static int getVersionCode() {
        if (versionCode < 0) {
            PackageManager pm = AppUtils.getApp().getPackageManager();
            try {
                PackageInfo packageInfo = pm.getPackageInfo(AppUtils.getApp().getPackageName(), 0);
                versionCode = packageInfo.versionCode;
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
        return versionCode;
    }

    public static String getPackageName() {
        if (TextUtils.isEmpty(packageName)) {
            PackageManager pm = AppUtils.getApp().getPackageManager();
            try {
                PackageInfo packageInfo = pm.getPackageInfo(AppUtils.getApp().getPackageName(), 0);
                packageName = packageInfo.packageName;
            } catch (PackageManager.NameNotFoundException e) {
                e.printStackTrace();
            }
        }
        if (packageName == null) {
            packageName = "";
        }
        return packageName;
    }

    /**
     * 获取应用程序名称
     */
    public static String getAppName() {
        int labelRes = getPackageInfo().applicationInfo.labelRes;
        return getApp().getResources().getString(labelRes);
    }

    /**
     * 获取App包 信息版本号
     */
    public static PackageInfo getPackageInfo() {
        PackageManager packageManager = getApp().getPackageManager();
        PackageInfo packageInfo = null;
        try {
            packageInfo = packageManager.getPackageInfo(getApp().getPackageName(), 0);
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
        return packageInfo;
    }


}

File: utilslib/src/main/java/com/ashlikun/utils/AppUtils.kt
package com.ashlikun.utils

import android.app.Application
import android.content.Context
import android.content.pm.PackageInfo
import android.content.res.Resources
import android.net.Uri
import android.os.Build
import android.view.View
import androidx.core.content.FileProvider
import com.ashlikun.utils.bug.BugUtils
import com.ashlikun.utils.provider.BaseContentProvider
import com.ashlikun.utils.ui.ActivityManager
import com.ashlikun.utils.ui.fActivity
import com.ashlikun.utils.ui.fCActivity
import com.ashlikun.utils.ui.resources.ResUtils
import java.io.File

/**
 * @author　　: 李坤
 * 创建时间: 2021.12.15 15:54
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：App 工具类
 */
/**
 * 作者　　: 李坤
 * 创建时间: 2022/10/13　10:35
 * 邮箱　　：496546144@qq.com
 *
 * 功能介绍：自定义view预览的问题
 */


object AppUtils {
    var isDebug = false
    lateinit var app: Application
        private set
    lateinit var base: Context
        private set

    /**
     * 编辑模式的Context
     */
    var editModeContext: Context? = null
    fun View.initAppEditMode() {
        if (isInEditMode) {
            AppUtils.base = context
            AppUtils.editModeContext = context
            AppUtils.defaultContextCall = {
                context
            }
        }
    }

    /**
     * 默认的Context
     * 为了适配使用Activity的,不是一股脑调用Application的
     * 当后台调用的时候可能不是调用者的Activity
     * 注意内存泄露
     */
    var defaultContextCall: () -> Context = {
//        (ActivityManager.foregroundActivity ?: app)
        app
    }

    fun attachBaseContext(context: Context?) {
        if (context != null) base = context
    }

    /**
     * 一定要在Application里面调用
     * 在全部库的最前面调用
     */
    fun init(myApp: Application, isDebug: Boolean = false) {
        this.isDebug = isDebug
        app = myApp
        base = myApp
        myApp.registerActivityLifecycleCallbacks(activityLifecycleCallbacks)
        //解决一些疑难杂症
        BugUtils.init()
        //监听FPS
        if (isDebug) {
            FPSFrameCallback.init()
        }
    }

    /**
     * 取版本code
     */
    val versionCode by lazy {
        packageInfo.versionCode
    }

    /**
     * 版本名称
     */
    val versionName by lazy {
        packageInfo.versionName.orEmpty()
    }

    /**
     * 包名
     */
    val packageName by lazy {
        app.packageName.orEmpty()
    }

    var activityLifecycleCallbacks = ApplicationListener()

    /**
     * 前后台监听
     */
    fun addAppRunChang(listener: OnChangForeground) {
        activityLifecycleCallbacks.addOnChangListener(listener)
    }

    /**
     * 前后台监听
     */
    fun removeAppRunChang(listener: OnChangForeground) {
        activityLifecycleCallbacks.removeOnChangListener(listener)
    }

    /**
     * 兼容高版本的fileprovider
     */
    fun getUri(file: File): Uri {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            val authority = app.packageName + ".Provider"
            FileProvider.getUriForFile(app, authority, file)
        } else {
            Uri.fromFile(file)
        }
    }

    /**
     * 获取应用程序名称
     */
    val appName: String
        get() {
            val labelRes = packageInfo?.applicationInfo?.labelRes ?: 0
            return ResUtils.getString(labelRes)
        }

    /**
     * 获取App包 信息版本号
     */
    val packageInfo: PackageInfo
        get() = app.packageManager.getPackageInfo(app.packageName, 0)

    /**
     * 获取其他App包 信息版本号
     */
    fun getPackageInfo(packageName: String = app.packageName, flag: Int = 0) = try {
        app.packageManager.getPackageInfo(packageName, flag)
    } catch (e: Exception) {
        null
    }


    /**
     * Appication 的 resources
     */
    val appResources: Resources
        get() = runCatching {
            app.resources
        }.getOrNull() ?: editModeContext?.resources!!

    val defaultContext: Context
        get() = defaultContextCall()
    val defaultResources: Resources
        get() = defaultContext.resources

    /**
     * 栈顶的Activity
     */
    val fContext: Context
        get() = runCatching {
            fActivity ?: app
        }.getOrNull() ?: editModeContext!!

}